engine/access/state_stream/backend/registers_async.go

```diff
@@ -0,0 +1,64 @@
+package backend
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/state_synchronization"
+	"github.com/onflow/flow-go/storage"
+	"go.uber.org/atomic"
+)
+
+// RegistersAsyncStore has the same basic structure as access/backend.ScriptExecutor
+// TODO: use this interface for backend_scripts.go
+type RegistersAsyncStore struct {
+	registerIndex storage.RegisterIndex
+	reporter      state_synchronization.IndexReporter
+	initialized   *atomic.Bool
+	init          sync.Once
+}
+
+func NewRegistersAsyncStore() *RegistersAsyncStore {
+	return &RegistersAsyncStore{
+		initialized: atomic.NewBool(false),
+	}
+}
+
+// InitDataAvailable follows the same pattern of backend.ScriptExecutor
+// This method can be called at any time after the RegistersAsyncStore object is created and
+// calls to GetRegisterValues will return a storage.ErrHeightNotIndexed,
+// since we can't disambiguate between the underlying store before bootstrapping or just simply being behind sync
+func (r *RegistersAsyncStore) InitDataAvailable(
+	indexReporter state_synchronization.IndexReporter,
+	registers storage.RegisterIndex,
+) {
+	r.init.Do(func() {
+		defer r.initialized.Store(true)
+		r.reporter = indexReporter
+		r.registerIndex = registers
+	})
+}
+
+// RegisterValues atomically gets the register values from the underlying storage.RegisterIndex
+// Expected errors:
+//   - storage.ErrHeightNotIndexed if the store is still bootstrapping or if the values at the height is not indexed yet
+func (r *RegistersAsyncStore) RegisterValues(ids flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	if !r.isDataAvailable(height) {
+		return nil, storage.ErrHeightNotIndexed
+	}
+	result := make([]flow.RegisterValue, len(ids))
+	for idx, regId := range ids {
```

Comment:
nit: by convention, use `i` for the iterator since `ids` is a slice.
```suggestion
	for i, regId := range ids {
```